fabric-textures-v0/src/main/java/net/fabricmc/fabric/api/client/texture/ColorSniffer.java

```diff
@@ -0,0 +1,52 @@
+package net.fabricmc.fabric.impl.client.texture;
+
+import net.fabricmc.fabric.mixin.client.texture.SpriteAccessor;
+import net.minecraft.client.texture.NativeImage;
+import net.minecraft.client.texture.Sprite;
+
+/**
+ * Utility class for averaging Sprite colors.
+ * You should call one of these methods!
+ */
+public class ColorSniffer {
+
+	/**
+	 * @param sprite The Sprite you want to get the average color from.
+	 * @return The average ABGR color of the provided Sprite.
+	 */
+	public static int getAverageSpriteColor(Sprite sprite) { return getAverageSpriteColor(sprite, 0); }
+
+	/**
+	 * @param sprite The Sprite you want to get the average color from.
+	 * @param frame The specific frame of the sprite, in case it has more than one.
+	 * @return The average ABGR color of the provided Sprite, during the provided frame.
+	 */
+	public static int getAverageSpriteColor(Sprite sprite, int frame) {
+		if (sprite == null) {
+			return 0xFFFF00FF;
+		}
+
+		NativeImage image = ((SpriteAccessor) sprite).getImages()[frame];
+		int pixelCount = 0;
+		long avgR = 0, avgG = 0, avgB = 0;
+		for (int y = 0; y < sprite.getHeight(); y++) {
+			for (int x = 0; x < sprite.getWidth(); x++) {
+				int c = image.getPixelRgba(x, y);
+				if (((c >> 24) & 0xFF) != 0x00) {
+					avgB += ((c >> 16) & 0xFF) * ((c >> 16) & 0xFF);
+					avgG += ((c >> 8) & 0xFF) * ((c >> 8) & 0xFF);
+					avgR += (c & 0xFF) * (c & 0xFF);
+					pixelCount++;
```

Comment:
This seems like sane logic to me, but I would like to see some unit tests

Reply:
do I look like I know what a "unit test" is?

my current implementations return 
```
[CHAT] RGB-averaged color: -9005622
[CHAT] HSB-averaged color: -10384956
[STDOUT]: testing getAverageSpriteColor(sprite, RGB): 
[STDOUT]: 1 : 15ms
[STDOUT]: 2 : 14ms
[STDOUT]: 3 : 14ms
[STDOUT]: 4 : 14ms
[STDOUT]: 5 : 14ms
[STDOUT]: 6 : 16ms
[STDOUT]: 7 : 14ms
[STDOUT]: 8 : 15ms
[STDOUT]: 9 : 16ms
[STDOUT]: 10: 14ms
[STDOUT]: testing getAverageSpriteColor(sprite, HSV): 
[STDOUT]: 1 : 533ms
[STDOUT]: 2 : 541ms
[STDOUT]: 3 : 556ms
[STDOUT]: 4 : 558ms
[STDOUT]: 5 : 525ms
[STDOUT]: 6 : 540ms
[STDOUT]: 7 : 534ms
[STDOUT]: 8 : 534ms
[STDOUT]: 9 : 572ms
[STDOUT]: 10: 563ms
``` while crunching the numbers for a 2048x2048 sprite.
you don't need to tell me that the second method looks unacceptably slow, but vanilla 16x16 sprites return 0ms through both methods. I'm looking into some way to minimize the performance impact from egregiously large sprites, preferably more elegant than just quartering the sample area and rolling with it

Reply:
I wasn't talking of performance specifically but to verify your color averaging logic will output the right values

Reply:
well...
1. rgb-wise, half the time what you think is "right" isn't what the `computer` considers "right". an rgb average is an rgb average, meaning blue pixel + yellow pixel = pus-colored (slightly greenish gray) pixel.
2. hsv-wise, this is the most accurate method of getting the average color per sprite... though I'm implementing logic to make it a lot less accurate on bigger sprites, as a means of not surrendering performance when given massive textures (let's be honest, >64px sprites have way too many gradients for accuracy to matter)

Reply:
Also can you share your hsv impl for us to look at so we know why it's slower?

Reply:
it's slower because there's `a lot` more math involved. the argb from all sampled pixels is converted into float[] h,s,b(v), then the values are rotated/averaged, and then the finished h,s,b is converted back into argb.
the commit is coming after I verify my half-hearted opt doesn't actually crash or error on certain edge-cases

Reply:
Really the mod devs should be smart and cache the colors for non-animated items after a resource reload so they don't have to average the color 144 times a second.

If you are smart about your mod the latency basically is negligible then

Reply:
this pr mainly exists because I'm tossing around with the topic in my mod. I'm actually using the primordial version of this code in the renderer (with caching, of course) and, in my opinion, it having **any** sort of noticeable latency would be unacceptable.